File: java/src/day18/LavaductLagoon.java
package day18;

import utils.Node;
import utils.Pair;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.*;

public class LavaductLagoon {
    private static List<Pair<String, Integer>> diggingInsPart1 = new ArrayList<>();
    private static List<String> diggingInsPart2 = new ArrayList<>();

    private static Map<String, List<Integer>> dig = Map.ofEntries(
            Map.entry("R", List.of(0, 1)),
            Map.entry("L", List.of(0, -1)),
            Map.entry("D", List.of(1, 0)),
            Map.entry("U", List.of(-1, 0))
    );

    private static Map<String, String> decode = Map.ofEntries(
            Map.entry("0", "R"),
            Map.entry("1", "D"),
            Map.entry("2", "L"),
            Map.entry("3", "U")
    );

    public static void main(String[] args) {
        try {
            getFile("inputs/day18/input.txt");
            System.out.println("Part 1: " + part1());
            System.out.println("Part 2: " + part2());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    private static long part2() {
        List<Pair<String, Integer>> digIns = new ArrayList<>();
        for (String ins : diggingInsPart2) {
            String diggDir = decode.get(ins.substring(ins.length() - 2, ins.length() - 1));
            Integer hexString = Integer.decode(ins.substring(1, ins.length() - 2));
            digIns.add(new Pair<>(diggDir, hexString));
        }
        return solution(digIns);
    }

    private static long part1() {
        return solution(diggingInsPart1);
    }

    private static long solution(List<Pair<String, Integer>> diggingInstructions) {
        List<Node> nodes = new ArrayList<>();
        nodes.add(new Node(0, 0));
        Long boundryPoints = 0L;
        for (Pair<String, Integer> pair : diggingInstructions) {
            Node node = nodes.getLast();
            boundryPoints += pair.getValue();
            List<Integer> move = dig.get(pair.getKey());
            int x = node.getX() + move.get(0) * pair.getValue();
            int y = node.getY() + move.get(1) * pair.getValue();
            nodes.add(new Node(x, y));
        }

        long result = 0;

        // Shoelace formula
        for (int i = 0; i < nodes.size() - 1; i++) {
            Node node1 = nodes.get(i);
            Node node2 = nodes.get(i + 1);

            result += (((long) node1.getX() * node2.getY()) - ((long) node1.getY() * node2.getX()));
        }
        // Pick's theorem
        Long i = Math.abs(result / 2) - boundryPoints / 2 + 1;
        return i + boundryPoints;
    }

    private static void getFile(String file) throws IOException {
        List<String> input = Files.readAllLines(Path.of(file));
        for (String row : input) {
            String[] rowSplit = row.strip().split(" ");
            diggingInsPart1.add(new Pair<>(rowSplit[0], Integer.valueOf(rowSplit[1])));
            diggingInsPart2.add(rowSplit[2]);
        }
    }
}

File: type-script/day18/lavaduct-lagoon.ts
import * as fs from "fs";
import { Position } from "../utils";

const f: string = fs.readFileSync("../inputs/day18/input.txt", { encoding: "utf8" });
const digPlanFull: string[][] = f.trim().split("\n").map(x => x.trim().split(/\s+/));
const digPlan: { dir: string, num: number }[] = digPlanFull.map(x => { return { dir: x[0], num: Number(x[1]) } });
const digPlan2: { dir: string, num: number }[] = digPlanFull.map(x => {
    const ins: string = x[2].replace("(#", "").replace(")", "");
    const dir: string = ins[ins.length - 1];
    const numHex: number = Number("0x" + ins.substring(0, ins.length - 1));
    return { dir: directionToDig(dir), num: numHex }
});


function solution(digPlan: { dir: string, num: number }[]): number {
    const nodes: Position[] = [];
    let prevNode: Position = new Position(0, 0);
    nodes.push(prevNode);
    let boundary: number = 0;
    for (const instruction of digPlan) {
        const newNode: Position = createNewNode(prevNode, instruction);
        prevNode = newNode;
        boundary += instruction.num
        nodes.push(prevNode);
    }

    let A = shoelaceFormula(nodes)
    return A + boundary / 2 + 1;
}


function shoelaceFormula(nodes: Position[]): number {
    let shoe: number = 0;
    for (let i = 0; i < nodes.length - 1; i++) {
        const node1: Position = nodes[i];
        const node2: Position = nodes[i + 1];

        shoe += (node1.i * node2.j) - (node2.i * node1.j);
    }
    return Math.abs(shoe / 2);
}


function createNewNode(prevNode: Position, instruction: { dir: string, num: number }): Position {
    let x = prevNode.i;
    let y = prevNode.j;
    if (instruction.dir === "R") {
        y += instruction.num;
    } else if (instruction.dir === "L") {
        y -= instruction.num;
    } else if (instruction.dir === "U") {
        x -= instruction.num;
    } else if (instruction.dir === "D") {
        x += instruction.num;
    }
    return new Position(x, y);
}


function directionToDig(dir: string): string {
    let x: string = ""
    switch (dir) {
        case "0":
            x = "R"
            break;
        case "1":
            x = "D";
            break;
        case "2":
            x = "L";
            break;
        case "3":
            x = "U";
            break;
    }
    return x;
}


console.log("Part 1:", solution(digPlan));
console.log("Part 2:", solution(digPlan2));
